src/include/gurt/shm_utils.h

```diff
@@ -376,4 +376,139 @@ shm_ht_rec_data(d_shm_ht_rec_loc_t rec_loc, int *err);
 int
 shm_ht_rec_num_ref(d_shm_ht_rec_loc_t rec_loc);
 
+/* dynamic allocation if data is larger than this threshold */
+#define LRU_ALLOC_SIZE_THRESHOLD (4096)
+
+/* Node of entry in LRU cache */
+typedef struct shm_lru_node {
+	/* key size*/
+	uint32_t    key_size;
+	/* data size*/
+	uint32_t    data_size;
+	/* store the offset to key */
+	long int    key;
+	/* store the offset to data */
+	long int    data;
+	/* the reference count of this record */
+	_Atomic int ref_count;
+	/* off_prev and off_next are used in doubly linked list for LRU */
+	int         off_prev;
+	int         off_next;
+	/* offset to the next node in hash chain in each bucket for allocated node. point to next
+	 * available node for free nodes
+	 */
+	int         off_hnext;
+	/* the index of hash bucket this record is in */
+	uint32_t    idx_bucket;
+	/* the index of sub-cache this record is in */
+	uint32_t    idx_subcache;
+} shm_lru_node_t;
+
+/* This implementation of shm LRU is mainly optimized for performance by using pre-allocated buffer
+ * when possible
+ */
+
+typedef struct {
+	/* number of nodes allocated in this shard */
+	uint32_t      size;
+	/* Most recently used node */
+	int           off_head;
+	/* Least recently used node */
+	int           off_tail;
+	/* First available/free node */
+	int           first_av;
+	/* the offset to the array of offset of hash buckets */
+	long int      off_hashbuckets;
+	/* the offset to the array of preallocated array of nodes */
+	long int      off_nodelist;
+	/* the offset to the array of preallocated array of keys */
+	long int      off_keylist;
+	/* the offset to the array of preallocated array of data */
+	long int      off_datalist;
+
+	d_shm_mutex_t lock;
+	char          pad[8];
+} shm_lru_cache_var_t;
+
+/* LRU Cache structure */
+typedef struct {
+	/* the number of sub-cache to use */
+	uint32_t n_subcache;
+	/* max number of nodes to hold in each shard */
+	uint32_t capacity;
+	/* the size of key. zero means key size is variable */
+	uint32_t key_size;
+	/* the size of data. zero means data size is variable */
+	uint32_t data_size;
+
+	/* the offset to shm_lru_cache_var_t */
+	long int off_cache_list;
+
+	char     pad[8];
+} shm_lru_cache_t;
+
+/**
+ * create LRU cache
+ *
+ * \param[in] n_subcache	the number of subcache. Each subcache owns a private lock
```

Comment:
can you explain what subcache means here? as a user of this API, i would have no clue what this means.

Reply:
Other than store all records in a large cache, we split records into multiple partitions. Each partition is called as a subcache and uses its own lock. The partition selection is based on the higher 32 bits of the hash value of the key. We could remove the argument here and always set n_subcache as the number of cores available on current client node to simplify the API.